File: lib/imagereader/nsiimagereader.cc
/* Image reader for Northstar floppy disk images */

#include "lib/core/globals.h"
#include "lib/config/flags.h"
#include "lib/data/sector.h"
#include "lib/imagereader/imagereader.h"
#include "lib/data/image.h"
#include "lib/core/logger.h"
#include "lib/imagereader/imagereader.pb.h"
#include <algorithm>
#include <iostream>
#include <fstream>

class NsiImageReader : public ImageReader
{
public:
    NsiImageReader(const ImageReaderProto& config): ImageReader(config) {}

    std::unique_ptr<Image> readImage() override
    {
        std::ifstream inputFile(
            _config.filename(), std::ios::in | std::ios::binary);
        if (!inputFile.is_open())
            error("cannot open input file");

        const auto begin = inputFile.tellg();
        inputFile.seekg(0, std::ios::end);
        const auto end = inputFile.tellg();
        const auto fsize = (end - begin);

        log("NSI: Autodetecting geometry based on file size: {}", fsize);

        unsigned numTracks = 35;
        unsigned numSectors = 10;
        unsigned numHeads = 2;
        unsigned sectorSize = 512;

        switch (fsize)
        {
            case 358400:
                numHeads = 2;
                sectorSize = 512;
                break;

            case 179200:
                numHeads = 1;
                sectorSize = 512;
                break;

            case 89600:
                numHeads = 1;
                sectorSize = 256;
                break;

            default:
                error("NSI: unknown file size");
        }

        size_t trackSize = numSectors * sectorSize;

        log("reading {} tracks, {} heads, {} sectors, {} bytes per sector, {} "
            "kB total",
            numTracks,
            numHeads,
            numSectors,
            sectorSize,
            numTracks * numHeads * trackSize / 1024);

        std::unique_ptr<Image> image(new Image);
        unsigned sectorFileOffset;

        for (unsigned head = 0; head < numHeads; head++)
        {
            for (unsigned track = 0; track < numTracks; track++)
            {
                for (unsigned sectorId = 0; sectorId < numSectors; sectorId++)
                {
                    if (head == 0)
                    { /* Head 0 is from track 0-34 */
                        sectorFileOffset =
                            track * trackSize + sectorId * sectorSize;
                    }
                    else
                    { /* Head 1 is from track 70-35 */
                        sectorFileOffset =
                            (trackSize * numTracks) + /* Skip over side 0 */
                            ((numTracks - track - 1) * trackSize) +
                            (sectorId * sectorSize); /* Sector offset from
                                                        beginning of track. */
                    }

                    inputFile.seekg(sectorFileOffset, std::ios::beg);

                    Bytes data(sectorSize);
                    inputFile.read((char*)data.begin(), sectorSize);

                    const auto& sector = image->put(track, head, sectorId);
                    sector->status = Sector::OK;
                    sector->data = data;
                }
            }
        }

        image->setGeometry({.numTracks = numTracks,
            .numSides = numHeads,
            .numSectors = numSectors,
            .sectorSize = sectorSize});
        return image;
    }
};

std::unique_ptr<ImageReader> ImageReader::createNsiImageReader(
    const ImageReaderProto& config)
{
    return std::unique_ptr<ImageReader>(new NsiImageReader(config));
}

File: java/com/cowlark/fluxengine/imagereader/NsiImageReader.java
package com.cowlark.fluxengine.imagereader;

import com.cowlark.fluxengine.core.ByteReader;
import com.cowlark.fluxengine.core.Bytes;
import com.cowlark.fluxengine.core.FluxEngineException;
import com.cowlark.fluxengine.core.Logger;
import com.cowlark.fluxengine.data.Geometry;
import com.cowlark.fluxengine.data.Image;
import com.cowlark.fluxengine.data.Sector;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/* Image reader for Northstar floppy disk images */
public class NsiImageReader extends ImageReader
{
    public NsiImageReader(ImageReaderProto config)
    {
        super(config);
    }

    @Override
    public Image readImage()
    {
        Bytes data;
        try
        {
            data = new Bytes(Files.readAllBytes(Path.of(config.getFilename())));
        } catch (IOException e)
        {
            throw new FluxEngineException("cannot open input file");
        }
        int fsize = data.size();

        Logger.logf("NSI: Autodetecting geometry based on file size: " + fsize);

        int numCylinders = 35;
        int numSectors = 10;
        int numHeads = 2;
        int sectorSize = 512;

        switch (fsize)
        {
            case 358400:
                numHeads = 2;
                sectorSize = 512;
                break;

            case 179200:
                numHeads = 1;
                sectorSize = 512;
                break;

            case 89600:
                numHeads = 1;
                sectorSize = 256;
                break;

            default:
                throw new FluxEngineException("NSI: unknown file size");
        }

        int trackSize = numSectors * sectorSize;

        Logger.logf("reading " + numCylinders + " tracks, " + numHeads + " heads, " + numSectors +
                " sectors, " + sectorSize + " bytes per sector, " +
                numCylinders * numHeads * trackSize / 1024 + " kB total");

        Image image = new Image();
        ByteReader br = new ByteReader(data);
        int sectorFileOffset;

        for (int head = 0; head < numHeads; head++)
        {
            for (int track = 0; track < numCylinders; track++)
            {
                for (int sectorId = 0; sectorId < numSectors; sectorId++)
                {
                    if (head == 0)
                    { /* Head 0 is from track 0-34 */
                        sectorFileOffset = track * trackSize + sectorId * sectorSize;
                    } else
                    { /* Head 1 is from track 70-35 */
                        sectorFileOffset = (trackSize * numCylinders) + /* Skip over side 0 */
                                ((numCylinders - track - 1) * trackSize) + (sectorId *
                                sectorSize); /* Sector offset from beginning of track. */
                    }

                    br.seek(sectorFileOffset);
                    Bytes sectorData = br.read(sectorSize);

                    Sector sector = image.put(track, head, sectorId);
                    sector.status = Sector.Status.OK;
                    sector.data = sectorData;
                }
            }
        }

        Geometry geometry = new Geometry();
        geometry.numCylinders = numCylinders;
        geometry.numHeads = numHeads;
        geometry.numSectors = numSectors;
        geometry.sectorSize = sectorSize;
        image.setGeometry(geometry);
        return image;
    }
}
